utility/src/main/java/org/oppia/util/parser/LiTagHandler.kt

```diff
@@ -1,37 +1,171 @@
 
 /**
- * [Html.TagHandler] implementation that processes <li> tags and creates bullets.
+ * Called when the HTML parser reaches an opening or closing tag.
+ * We only handle list tags and ignore other tags.
  *
- * Reference: https://github.com/davidbilik/bullet-span-sample
+ * <ul> and <ol> tags are pushed to the [lists] stack and popped when the closing tag is reached.
+ *
+ * <li> tags are handled by the [ListTag] instance corresponding to the parent tag.
+ *
+ * Reference: https://github.com/daphliu/android-spannable-list-sample/tree/master/app/src/main/java/com/daphneliu/sample/listspansample
  */
-class LiTagHandler : Html.TagHandler {
+class LiTagHandler(private val context: Context) : Html.TagHandler {
+
+  private val lists = Stack<ListTag>()
+
+  override fun handleTag(opening: Boolean, tag: String, output: Editable, xmlReader: XMLReader) {
+    when (tag) {
+      UL_TAG -> if (opening) {
+        // handle <ul>
+        lists.push(Ul())
+      } else {
+        // handle </ul>
+        lists.pop()
+      }
+      OL_TAG -> if (opening) {
+        // handle <ol>
+        lists.push(Ol())
+      } else {
+        // handle </ol>
+        lists.pop()
+      }
+      LI_TAG -> if (opening) {
+        // handle <li>
+        lists.peek().openItem(output)
+      } else {
+        // handle </li>
+        lists.peek().closeItem(output, indentation = lists.size - 1)
+      }
+    }
+  }
+
   /**
-   * Helper marker class. Based on [Html.fromHtml] implementation.
+   * Handler for <li> tags. Subclasses set the bullet appearance.
    */
-  class Bullet
+  private interface ListTag {
 
-  override fun handleTag(opening: Boolean, tag: String, output: Editable, xmlReader: XMLReader) {
-    if (tag == "li" && opening) {
-      output.setSpan(Bullet(), output.length, output.length, Spannable.SPAN_INCLUSIVE_EXCLUSIVE)
-    }
-    if (tag == "li" && !opening) {
-      output.append("<br>")
-      val lastMark = output.getSpans(0, output.length, Bullet::class.java).lastOrNull()
-      lastMark?.let {
-        val start = output.getSpanStart(it)
-        output.removeSpan(it)
-        if (start != output.length) {
-          output.setSpan(BulletSpan(), start, output.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
-        }
+    /**
+     * Called when an opening <li> tag is encountered.
+     *
+     * Inserts an invisible [Mark] span that doesn't do any styling.
+     * Instead, [closeItem] will later find the location of this span so it knows where the opening tag was.
+     */
+    fun openItem(text: Editable)
+
+    /**
+     * Called when a closing </li> tag is encountered.
+     *
+     * Pops out the invisible [Mark] span and uses it to get the opening tag location.
+     * Then, sets a [LeadingMarginSpan] from the opening tag position to closing tag position.
+     */
+    fun closeItem(text: Editable, indentation: Int)
+  }
+
+  /**
+   * Subclass of [ListTag] for unordered lists.
+   */
+  private inner class Ul : ListTag {
```

Comment:
Here & elsewhere: I suggest naming it out instead of abbreviating, e.g. ``UnorderedListTag``.

Reply:
Done.